### DISCUSSION TITLE
Any practical examples of using `d.multicall.filtered` on the command line?

### DISCUSSION THREAD

**Discussion by Rouneq** [rakshasa/rtorrent#1275] [Q&A] [answered]
rtorrent 0.9.8
xmlrpc-c command-line utility

Title says it all. Trying to find a practical example that works with base rtorrent to return a set of data specific to a single torrent. I can issue multiple calls for this data (currently doing so in my scripts). I know `d.multicall2` will return the data for all torrents in the specified view. According to the [0.9.8 release info](https://github.com/rakshasa/rtorrent/releases/tag/v0.9.8) `d.multicall.filtered` is available.

Unfortunately, the examples to filter all rely on rtorrent-PS rather than base rtorrent. Any guidance out there on how to use this other than what is in the [docs](https://rtorrent-docs.readthedocs.io/en/latest/cmd-ref.html#term-d-multicall-filtered)?

**Comment by kannibalox** [comment]
The usage between rtorrent-PS and base rtorrent is the same. I highly recommend using `rtxmlrpc` over the `xmlrpc-c` tool for flexibility, but the calls look pretty much the same. This example uses d.multicall.filtered to grab the name and directory of a specific torrent:
```bash
# xmlrpc-c
xmlrpc https://rtorrent-hostname/RPC2 d.multicall.filtered s/'' default 'equal=d.hash=\,cat=1078A67BBAA9184466BD551BB4460A1D9062A0AA' d.hash= d.name=
# rtxmlrpc
rtxmlrpc -U https://rtorrent-hostname/RPC2 d.multicall.filtered '' default 'equal=d.hash=,cat=1078A67BBAA9184466BD551BB4460A1D9062A0AA' d.hash= d.name=
```

Note that you can also do something similar with the built-in `system.multicall`, but calling that from the shell is a little tortuous, I'd recommend it more if you're trying to accomplish something programmatically.

**Reply by Rouneq** [reply]
Thanks for the feedback @kannibalox . I've taken a look at `rtxmlrpc` (and the broader `pyrocore`), but didn't feel comfortable enough with it to give it a go. Perhaps one day.

Using your example for `xmlrpc` substituting in my own rtorrent host and a local hash code just resulted in returning all my active torrents rather than just the single filtered item. This is what I've encountered previously. It's either all or none in my testing. This is what lead me to think I wasn't using it correctly (which I'm not), but if the results are the same then I'm still missing something.

```
[server ~/private/rtorrent] xmlrpc https://<server-host>/rtorrent/rpc -username=<username> -password=<password> d.multicall.filtered '' default 'equal=d.hash=\,cat=15CA98BAE36F9C315E74D437A03DB6608244DF95' d.hash= d.name=
Result:

Array of 1334 items:
  Index  0 Array of 2 items:
             Index  0 String: '582F6EF055CC59EA671A6B69B2BEA1DEDC820F56'
             Index  1 String: 'Awesome_Torrent_Name'
   .
   .
   .
  Index 1109 Array of 2 items:
               Index  0 String: '15CA98BAE36F9C315E74D437A03DB6608244DF95'
               Index  1 String: 'Fabulous_Torrent_Name'
   .
   .
   .
  Index 1333 Array of 2 items:
               Index  0 String: '641446F2AC66709DCF2E37033D21DE9C56D1E784'
               Index  1 String: 'Stupendous_Torrent_Name'
[server ~/private/rtorrent]
```

**Reply by kannibalox** [reply, marked as answer]
Aha, I'd have to figure out exactly why, but it turns out vanilla rtorrent requires an extra command that -PS and other forks don't. This works for me:

```bash
xmlrpc https://rtorrent-hostname/RPC2 d.multicall.filtered s/'' default 'equal=cat=$d.hash=\,cat=1078A67BBAA9184466BD551BB4460A1D9062A0AA' d.hash= d.name=
```

Also, if you don't want to go through the whole pyrocore setup, I made an fork that works on python 3 and only uses pip for installation of a single package: https://kannibalox.github.io/pyrosimple/

**Reply by Rouneq** [reply]
That got me close. For some reason, it didn't like the backslash.

```
xmlrpc https://rtorrent-hostname/RPC2 d.multicall.filtered s/'' default 'equal=cat=$d.hash=,cat=1078A67BBAA9184466BD551BB4460A1D9062A0AA' d.hash= d.name=
```

successfully filtered down to just the specified torrent.

Thanks a bunch for the assist.

**Reply by Rouneq** [reply]
Just a quick follow-up: Where did you pull the `equal` syntax from? I'm sure there are other filters I'll be interested in doing. For example, finding all torrents where the download rate is greater than zero (or not equal zero).

**Reply by kannibalox** [reply]
I'll give one final plug for the pyro tools, which already has that kind of logic built-in:
```bash
rtcontrol 'down>0' -o hash,name
```
The `--debug` flag will even show you the filtered command it built:
```
DEBUG:pyrosimple.util.rpc:method 'd.multicall.filtered', params ('', 'default', 'greater=d.down.rate=,value=0', 'd.down.rate=', 'd.hash=', 'd.name=')
```

The list of conditional commands is here: https://rtorrent-docs.readthedocs.io/en/latest/cmd-ref.html#conditional-operators, and while that's technically the extent of the syntax, actually using it is obviously not the most intuitive process. Most of my understanding just comes from trial and error when helping other people out.
